### PR TITLE
fix(runtime): add compressed {&packpath}/start/*/pack/*[/after] representation to &rtp

### PR DESCRIPTION
**Pull request by bfredl**
by suggestion by @tpope https://github.com/neovim/neovim/pull/15632#issuecomment-930639395

Summary:

We can add `XDG_DATA_DIR/nvim/site/pack/*/start/*` (et al) as an _unexpanded_ wildchar to `&rtp` which keeps it both short and explicit and still supporting `globpath(&rtp, ...)`.

ref #15101

### CODE REVIEW

File: test/functional/core/startup_spec.lua [hunk 1]
```diff
@@ -354,12 +354,34 @@ describe('startup', function()
...
+  it("handles the correct order with opt packages and globpath(&rtp, ...)", function()
+    pack_clear [[ set loadplugins | lua _G.test_loadorder = {} ]]
+    command [[
+      packadd! superspecial
+      for x in globpath(&rtp, "filen.lua",1,1)
+        call v:lua.dofile(x)
+      endfor
+    ]]
+    eq({'ordinary', 'SuperSpecial', 'FANCY', 'mittel', 'SuperSpecial after', 'FANCY after', 'ordinary after'}, exec_lua [[ return _G.test_loadorder ]])
+  end)
```
Comment (justinmk) [inline test/functional/core/startup_spec.lua:382]: do we want a test that checks for the presence of the glob string in `&rtp`?

File: src/nvim/runtime.c [hunk 2]
```diff
@@ -521,7 +540,7 @@ static void source_all_matches(char_u *pat)
 }
 
 /// Add the package directory to 'runtimepath'
-static int add_pack_dir_to_rtp(char_u *fname)
+static int add_pack_dir_to_rtp(char_u *fname, bool is_pack)
```
Comment (justinmk) [inline src/nvim/runtime.c:543]: does the docstring need to be updated?

**Review by justinmk — APPROVED** [review]

File: src/nvim/runtime.c [hunk 3]
```diff
@@ -332,40 +332,55 @@ int do_in_path_and_pp(char_u *path, char_u *name, int flags, DoInRuntimepathCB c
...
+  }
+}
+
+static void expand_pack_entry(RuntimeSearchPath *search_path, Map(String, handle_T) *rtp_used,
+                              CharVec *after_path, char_u *pack_entry)
+{
+  static char buf[MAXPATHL];
+  char *(start_pat[]) = { "/pack/*/start/*", "/start/*" };  // NOLINT
+  for (int i = 0; i < 2; i++) {
+    if (!(STRLEN(pack_entry) + STRLEN(start_pat[i]) + 1 < MAXPATHL)) {
```
Comment (justinmk) [inline src/nvim/runtime.c:373]: nit: this negative logic (used in 3 places) could be more readable if the `!(... < ...)` is switched to `... >= ...`

**Review by justinmk — APPROVED** [review]

**Comment by tpope** [comment]
Can confirm this stops the bleeding.

**Pull request merged**